examples/step-66/step-66.cc

```diff
@@ -0,0 +1,1156 @@
+                                additional_data);
+
+      system_matrix.initialize(system_mf_storage);
+      residual_operator.initialize(system_mf_storage);
+    }
+
+    system_matrix.initialize_dof_vector(solution);
+    system_matrix.initialize_dof_vector(newton_update);
+    system_matrix.initialize_dof_vector(system_rhs);
+
+
+    const unsigned int nlevels = triangulation.n_global_levels();
+    mg_matrices.resize(0, nlevels - 1);
+    mg_solution.resize(0, nlevels - 1);
+
+    std::set<types::boundary_id> dirichlet_boundary;
+    dirichlet_boundary.insert(0);
+    mg_constrained_dofs.initialize(dof_handler);
+    mg_constrained_dofs.make_zero_boundary_constraints(dof_handler,
+                                                       dirichlet_boundary);
+
+    for (unsigned int level = 0; level < nlevels; ++level)
+      {
+        IndexSet relevant_dofs;
+        DoFTools::extract_locally_relevant_level_dofs(dof_handler,
+                                                      level,
+                                                      relevant_dofs);
+
+        ConstraintMatrix level_constraints;
+        level_constraints.reinit(relevant_dofs);
+        level_constraints.add_lines(
+          mg_constrained_dofs.get_boundary_indices(level));
+        level_constraints.close();
+
+        typename MatrixFree<dim, float>::AdditionalData additional_data;
+        additional_data.tasks_parallel_scheme =
+          MatrixFree<dim, float>::AdditionalData::partition_color;
+        additional_data.mapping_update_flags =
+          (update_values | update_gradients | update_JxW_values |
+           update_quadrature_points);
+        additional_data.level_mg_handler = level;
+        std::shared_ptr<MatrixFree<dim, float>> mg_mf_storage_level(
+          new MatrixFree<dim, float>());
+        mg_mf_storage_level->reinit(dof_handler,
+                                    level_constraints,
+                                    QGauss<1>(fe.degree + 1),
+                                    additional_data);
+
+        mg_matrices[level].initialize(mg_mf_storage_level,
+                                      mg_constrained_dofs,
+                                      level);
+        mg_matrices[level].initialize_dof_vector(mg_solution[level]);
+      }
+  }
+
+
+
+  // @sect4{GelfandProblem::assemble_rhs}
+
+  // Using the implementation of the ResidualOperator the assembly of the right
+  // hand side becomes now a very easy task. We just call the main function of
+  // the residual operator and multiply the result with minus one.
+  //
+  // Experiences show that using the cell_loop of the MatrixFree class together
+  // with the FEEvaluation class is much faster than a classical implementation.
+  template <int dim>
+  void GelfandProblem<dim>::assemble_rhs()
+  {
+    TimerOutput::Scope t(computing_timer, "assemble right hand side");
+
+    residual_operator.apply(system_rhs, solution);
```

Comment:
Yes, moving the code with a functor here would be very nice.

Reply:
Just implemented a version, which is now even faster as with the ResidualOperator approach.